### PR TITLE
webapp: add support for using new OneDeploy API for deployments  in CLI

### PR DESCRIPTION
**Pull request by dannysongg**
- Allows users to deploy various types of artifacts such as app.war, app.jar, static files, and zip files
- Supports asynchronous deployments, and deployments to a custom path.
- Deployment type is inferred from the extension of the file being deployed. This can be overridden using the --type parameter.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [X] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [X] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [X] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your PR is merged into master branch, a new PR will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repo and upgrade the version in the PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
webapp

File: src/webapp/azext_webapp/_help.py [hunk 1]
```diff
@@ -59,3 +59,8 @@
         - name: Upload files from the current directory to an Azure Container Registry, then build a container image and deploy it to a web app. The Azure Container Registry must already exist.
           text: az webapp container up -n AppName --registry-rg ContainerRegistryResourceGroup --registry-name ContainerRegistryName
 """
+
+helps['webapp deploy'] = """
+    type: command
+    short-summary: Deploys a given artifact to Azure Web Apps through OneDeploy.
```
Comment (panchagnula) [inline src/webapp/azext_webapp/_help.py:65]: isn't this going to be scoped only for certain extensions currently? Like .war, .zip so would be good to be clear. Please check these help content with your PM as well

File: src/webapp/azext_webapp/custom.py [hunk 2]
```diff
@@ -308,3 +310,75 @@ def _should_create_new_asp(cmd, rg_name, asp_name, location):
...
+    # check if there's an ongoing process
+    if res.status_code == 409:
+        raise CLIError("There may be an ongoing deployment or your app setting has WEBSITE_RUN_FROM_PACKAGE. "
+                       "Please track your deployment in {} and ensure the WEBSITE_RUN_FROM_PACKAGE app setting "
+                       "is removed.".format(deployment_status_url))
+
+    # check the status of async deployment
+    if res.status_code == 202:
+        logger.warning("Asynchronous deployment request has been recieved")
+        response = _check_zip_deployment_status(cmd, resource_group_name, name, deployment_status_url, authorization, timeout)
```
Comment (panchagnula) [inline src/webapp/azext_webapp/custom.py:381]: Have verified that this check deployment status works here as well?

File: src/webapp/azext_webapp/__init__.py [hunk 3]
```diff
@@ -25,6 +25,7 @@ def load_command_table(self, _):
         with self.command_group('webapp') as g:
             g.custom_command('container up', 'create_deploy_container_app', exception_handler=ex_handler_factory())
             g.custom_command('remote-connection create', 'create_tunnel')
+            g.custom_command('deploy', 'enable_one_deploy')
```
Comment (panchagnula) [inline src/webapp/azext_webapp/__init__.py:28]: We need the arguments defined for this command inside load_arguments below.

File: src/webapp/azext_webapp/_help.py (hunk 1 above)
Comment (dannysongg) [reply src/webapp/azext_webapp/_help.py:65]: The OneDeploy API can deploy any extension to a webapp as a static file. However it does take specific actions for certain extensions such as .war, .jar, and .zip.

File: src/webapp/azext_webapp/custom.py (hunk 2 above)
Comment (dannysongg) [reply src/webapp/azext_webapp/custom.py:381]: Check deployment status has the same functionality as it does when calling 'az webapp deployment source config-zip'

File: src/webapp/azext_webapp/custom.py [hunk 4]
```diff
@@ -308,3 +310,75 @@ def _should_create_new_asp(cmd, rg_name, asp_name, location):
...
+    headers['Cache-Control'] = 'no-cache'
+    headers['User-Agent'] = get_az_user_agent()
+
+    import requests
+    import os
+
+    # Read file content
+    with open(os.path.realpath(os.path.expanduser(src)), 'rb') as fs:
+        artifact_content = fs.read()
+        logger.warning("Starting deployment...")
```
Comment (qwordy) [inline src/webapp/azext_webapp/custom.py:365]: It is more like an info not a warning

File: src/webapp/azext_webapp/custom.py [hunk 5]
```diff
@@ -308,3 +310,75 @@ def _should_create_new_asp(cmd, rg_name, asp_name, location):
...
+    if res.status_code == 409:
+        raise CLIError("There may be an ongoing deployment or your app setting has WEBSITE_RUN_FROM_PACKAGE. "
+                       "Please track your deployment in {} and ensure the WEBSITE_RUN_FROM_PACKAGE app setting "
+                       "is removed.".format(deployment_status_url))
+
+    # check the status of async deployment
+    if res.status_code == 202:
+        logger.warning("Asynchronous deployment request has been recieved")
+        response = _check_zip_deployment_status(cmd, resource_group_name, name, deployment_status_url, authorization, timeout)
+        return response
```
Comment (qwordy) [inline src/webapp/azext_webapp/custom.py:382]: Is there any other status code that should be handled? And how to handle exception like connection error?

File: src/webapp/azext_webapp/custom.py (hunk 4 above)
Comment (dannysongg) [reply src/webapp/azext_webapp/custom.py:365]: Other commands use logger.warning to print information to the terminal. I used here to remain consistent with the other code.

File: src/webapp/azext_webapp/custom.py (hunk 5 above)
Comment (dannysongg) [reply src/webapp/azext_webapp/custom.py:382]: The code currently handles all possible status codes the can be returned during a OneDeploy deployment. I believe that other errors such as connection errors will be handled externally. Please correct me if I am incorrect about this.

**Comment by yonzhan** [comment]
webapp

File: src/webapp/azext_webapp/_help.py [hunk 6]
```diff
@@ -59,3 +59,8 @@
         - name: Upload files from the current directory to an Azure Container Registry, then build a container image and deploy it to a web app. The Azure Container Registry must already exist.
           text: az webapp container up -n AppName --registry-rg ContainerRegistryResourceGroup --registry-name ContainerRegistryName
 """
+
+helps['webapp deploy'] = """
+    type: command
```
Comment (panchagnula) [inline src/webapp/azext_webapp/_help.py:64]: Can you add some examples here for different types and if the path can be relative vs absolute path?

File: src/webapp/azext_webapp/custom.py [hunk 7]
```diff
@@ -308,3 +310,75 @@ def _should_create_new_asp(cmd, rg_name, asp_name, location):
                 item.location == location):
             return False
     return True
+
+
+def enable_one_deploy(cmd, resource_group_name, name, src, deploy_type=None, is_async=None, target_path=None, timeout=None, slot=None):
+    import ntpath
+    logger.warning("Preparing for deployment")
```
Comment (panchagnula) [inline src/webapp/azext_webapp/custom.py:317]: can you make this logger.info() so the logs are minimal

File: src/webapp/azext_webapp/custom.py [hunk 8]
```diff
@@ -308,3 +310,75 @@ def _should_create_new_asp(cmd, rg_name, asp_name, location):
...
+    user_name, password = _get_site_credential(cmd.cli_ctx, resource_group_name, name, slot)
+
+    try:
+        scm_url = _get_scm_url(cmd, resource_group_name, name, slot)
+    except ValueError:
+        raise CLIError('Failed to fetch scm url for function app')
```
Comment (panchagnula) [inline src/webapp/azext_webapp/custom.py:323]: why does this say function app here? can you update as failed to fetch sam url for app {} -> show app name as well.

File: src/webapp/azext_webapp/__init__.py [hunk 9]
```diff
@@ -51,14 +52,26 @@ def load_arguments(self, _):
             c.argument('dryrun', help="show summary of the create and deploy operation instead of executing it", default=False, action='store_true')
             c.argument('registry_rg', help="the resource group of the Azure Container Registry")
             c.argument('registry_name', help="the name of the Azure Container Registry")
+            c.argument('slot', help="the deployment slot")
```
Comment (qwordy) [inline src/webapp/azext_webapp/__init__.py:55]: Maybe you need to make the help messages consistent.
In the other two places, you use
```
Name of the deployment slot to use
```

**Pull request merged**